js/main.js

```diff
@@ -355,94 +372,99 @@ class App extends React.Component {
             );
           }
         }
-        this.setState({
+        setStoreMeta((prev) => ({
+          ...prev,
           envList: cmd.data,
           layoutLists: layoutLists,
-        });
+        }));
         break;
       case 'layout_update':
-        this.parseLayoutsFromServer(cmd.data);
+        parseLayoutsFromServer(cmd.data);
         break;
       default:
         console.error('unrecognized command', cmd);
     }
   };
 
-  disconnect = () => {
-    this._socket.close();
+  // we need to update the socket-callback so that we have an up-to date state
+  if (_socket.current) _socket.current.onmessage = _handleMessage;
```

Comment:
> > `consistent_pane_copy `
> 
> I'll be honest, I've forgotten what the purpose of this initially was

If I understood correctly, the variable exists, because all functions called by `socket.onhandler` or below it in the calling hierarchy use the state of the initialization of the socket itself.
On the other hand, all other functions exist in subsequent calls of the outer react function and thus have updated states.
To tackle this,
- the variable `consistent_pane_copy` is the one variable that ensures to always contain the correct state of the server, implicitly telling the newer render-functions-instances about all changes
- `_checkWindow` for instance applies changes always to both, the current state pane store and the `consistent_pane_copy` to ensure both are synced in the "long term"
- the old code had sometimes overwritten state variables using operations such as `state.panes = newPanes` to update variables in the short term (i.e. in case of race conditions)

Migrating that part in a `1:1` way was the hardest part and I could be completely wrong about what the variable actually does.
However, I've added here an update of the socket to always use "current" `_handleMessage` from the current render-call. This one line has fixed several problems I had during this PR.

Reply:
Interestingly, in the new implementation I could remove the variable completely without any failing tests.
(See #901).